ephios/plugins/federation/views/api.py

```diff
@@ -0,0 +1,125 @@
+from urllib.parse import urljoin
+
+import requests
+from django.conf import settings
+from django.shortcuts import get_object_or_404, redirect
+from django.urls import reverse
+from django.views import View
+from oauth2_provider.contrib.rest_framework import TokenHasScope
+from oauthlib.oauth2 import WebApplicationClient
+from requests_oauthlib import OAuth2Session
+from rest_framework.exceptions import PermissionDenied
+from rest_framework.generics import CreateAPIView, ListAPIView
+from rest_framework.permissions import AllowAny
+
+from ephios.core.models import Event, Qualification
+from ephios.plugins.federation.models import FederatedGuest, FederatedUser
+from ephios.plugins.federation.serializer import (
+    FederatedGuestCreateSerializer,
+    SharedEventSerializer,
+)
+
+
+class RedeemInviteCodeView(CreateAPIView):
+    serializer_class = FederatedGuestCreateSerializer
+    queryset = FederatedGuest.objects.all()
+    authentication_classes = []
+    permission_classes = [AllowAny]
+
+
+class SharedEventListView(ListAPIView):
+    serializer_class = SharedEventSerializer
+    permission_classes = [TokenHasScope]
+    required_scopes = []
+
+    def get_queryset(self):
+        try:
+            guest = self.request.auth.federatedguest_set.get()
+        except FederatedGuest.DoesNotExist as exc:
+            raise PermissionDenied from exc
+        return Event.objects.filter(federatedeventshare__shared_with=guest)
+
+
+class FederationOAuthView(View):
+    def get(self, request, *args, **kwargs):
+        if "code" not in request.GET.keys():
+            try:
+                return redirect(self._get_authorization_url())
+            except (KeyError, FederatedGuest.DoesNotExist) as exc:
+                raise PermissionDenied from exc
+        else:
+            self._oauth_callback()
+            return redirect("federation:event_detail", pk=self.request.session.pop("event"))
+
+    def _get_authorization_url(self):
+        guest = (
+            FederatedGuest.objects.get(pk=self.request.session["guest"])
+            if "guest" in self.request.session.keys()
+            else FederatedGuest.objects.get(url=self.request.GET["referrer"])
```

Comment:
these lines assume just on FederatedGuest per url and I was able to break that while testing around...

![image](https://github.com/ephios-dev/ephios/assets/5970076/7f6e2e07-0dbb-45ab-adde-5c30bbb18c61)

Reply:
No hard feelings but there are lots of places esp in this file where I think the code makes a lot of assumptions that might be broken with real users fiddeling around with this feature and I really don't want them to cause server errors.